teeracle/src/lib.rs

```diff
@@ -93,52 +114,91 @@ pub mod pallet {
-		pub fn remove_from_whitelist(origin: OriginFor<T>, mrenclave: [u8; 32]) -> DispatchResult {
+		pub fn remove_from_whitelist(
+			origin: OriginFor<T>,
+			data_source: MarketDataSourceString,
+			mrenclave: [u8; 32],
+		) -> DispatchResult {
 			ensure_root(origin)?;
-			ensure!(Self::is_whitelisted(mrenclave), <Error<T>>::ReleaseNotWhitelisted);
-			Whitelist::<T>::mutate(|mrenclaves| mrenclaves.retain(|m| *m != mrenclave));
-			Self::deposit_event(Event::RemovedFromWhitelist(mrenclave));
+			ensure!(
+				Self::is_whitelisted(data_source.clone(), mrenclave),
+				<Error<T>>::ReleaseNotWhitelisted
+			);
+			<Whitelists<T>>::mutate(data_source.clone(), |mrenclave_vec| {
+				mrenclave_vec.retain(|m| *m != mrenclave)
+			});
+			Self::deposit_event(Event::RemovedFromWhitelist(data_source, mrenclave));
 			Ok(())
 		}
 
 		#[pallet::weight(<T as Config>::WeightInfo::update_exchange_rate())]
 		pub fn update_exchange_rate(
 			origin: OriginFor<T>,
-			currency: CurrencyString,
+			data_source: MarketDataSourceString,
+			trading_pair: TradingPairString,
 			new_value: Option<ExchangeRate>,
 		) -> DispatchResultWithPostInfo {
 			let sender = ensure_signed(origin)?;
 			<pallet_teerex::Module<T>>::is_registered_enclave(&sender)?;
 			let sender_index = <pallet_teerex::Module<T>>::enclave_index(sender);
 			ensure!(
-				Self::is_whitelisted(<pallet_teerex::Module<T>>::enclave(sender_index).mr_enclave),
+				trading_pair.len() <= MAX_TRADING_PAIR_LEN,
+				Error::<T>::TradingPairStringTooLong
+			);
+			ensure!(
+				Self::is_whitelisted(
+					data_source.clone(),
+					<pallet_teerex::Module<T>>::enclave(sender_index).mr_enclave
+				),
 				<Error<T>>::ReleaseNotWhitelisted
 			);
 			if new_value.is_none() || new_value == Some(U32F32::from_num(0)) {
 				log::info!("Delete exchange rate : {:?}", new_value);
-				ExchangeRates::<T>::mutate_exists(currency.clone(), |rate| *rate = None);
-				Self::deposit_event(Event::ExchangeRateDeleted(currency));
+				ExchangeRates::<T>::mutate_exists(
+					trading_pair.clone(),
+					data_source.clone(),
+					|rate| *rate = None,
+				);
+				Self::deposit_event(Event::ExchangeRateDeleted(data_source, trading_pair));
 			} else {
 				log::info!("Update exchange rate : {:?}", new_value);
-				ExchangeRates::<T>::mutate_exists(currency.clone(), |rate| *rate = new_value);
-				Self::deposit_event(Event::ExchangeRateUpdated(currency, new_value));
+				ExchangeRates::<T>::mutate_exists(
+					trading_pair.clone(),
+					data_source.clone(),
+					|rate| *rate = new_value,
+				);
+				Self::deposit_event(Event::ExchangeRateUpdated(
+					data_source,
+					trading_pair,
+					new_value,
+				));
 			}
 			Ok(().into())
 		}
 	}
 }
 impl<T: Config> Pallet<T> {
-	fn is_whitelisted(mrenclave: [u8; 32]) -> bool {
-		Self::whitelist().contains(&mrenclave)
+	fn is_whitelisted(data_source: MarketDataSourceString, mrenclave: [u8; 32]) -> bool {
+		Self::whitelist(data_source).contains(&mrenclave)
```

Comment:
Here you should also be able to use `&data_source` and thereby pass `&MarketDataSourceString`. This will spare you a bunch of clones.